exhale/graph.py

```diff
@@ -2401,21 +2414,40 @@ class view hierarchy (<a href="..."> for the ``createTreeView = True`` option).
             unique_id = node.refid
 
             # special treatment for templates
-            first_lt = node.name.find("<")
-            last_gt  = node.name.rfind(">")
+            last_lt = node.name.rfind("<")
+            last_gt = node.name.rfind(">")
             # dealing with a template when this is true
-            if first_lt > -1 and last_gt > -1:
+            if last_lt > -1 and last_gt > -1:
                 # NOTE: this has to happen for partial / full template specializations
                 #       When specializations occur, the "<param1, param2>" etc show up
                 #       in `node.name`.
                 template_special = True
+                try:
+                    groups = utils.groupsFromBalancedBrackets(node.name, '<', '>')
+                    templates = []
+                    utils.groupsToNamedGroups(groups, templates)
+                    if type(templates[-1]) is type(str):  # it isn't a parameter list.
+                        title = templates[-1].split('::')[-1]
+                    else:
+                        template = utils.templateListToNodeName(templates)
+                        title = template.split('::')[-1]
```

Comment:
@florianhumblot can you help me understand this code?  I don't think `templateListToNodeName` works correctly, I have an alternative implementation beneath this -- `groupsFromBalancedBrackets` and `groupsToNamedGroups` get folded into one, it should more or less produce the same thing as the alternate `utils.tokenize_template` but the new implementation also trims whitespace.

Right now if you run `tox -e py -- -k test_html_output -s` you'll get dropped into the debugger, the print out for `title` is from your code and the `my_title` is from the new proposal:

```py
ipdb> title
'Base<2>'
ipdb> my_title
'Base< 2 >'
# ... continue ...
ipdb> title
'ImageBuffer<Image<1920, 1080>float, 128>'
ipdb> my_title
'ImageBuffer< Image< 1920, 1080 >, float, 128 >'
# ... continue ...
ipdb> title
'ImageBuffer<TImage, Tdata_t, 4>'
ipdb> my_title
'ImageBuffer< TImage, Tdata_t, 4 >'
# ... continue ...
ipdb> title
'InnerTemplatedStruct<4, dont_use_this>'
ipdb> my_title
'InnerTemplatedStruct< 4, dont_use_this >'
```

There seems to be a slight issue with `templateListToNodeName`, but when I was looking at the larger picture the question is -- do we even need this function?  AFAICT what's supposed to happen is the tokenized list of templates gets re-joined recursively, but this is ultimately building out the original `node.name` right?  If we can avoid needing to fix the function and instead just split the original `node.name` then I think we're good.

My understanding is the core change here is tokenizing the list, and if the last one in the template list is a string we have to grab that?

Sorry if that doesn't make sense, this one is kind of tricky!

Reply:
Ugh no I need to sleep on this -- from what I'm seeing, we always want `node.name.split("::")[-1]` (same as the `else`, and ... no special treatment)?  Sorry... I think I've lost sight of what's actually the problem and/or the test cases we created aren't adequate

Reply:
Right. We can't use `node.name` here, and I need to add a better test case. Consider the nested template class whose template parameter list has `::` in it.

So tokenizing needs to happen, but a little more care on rebuilding the last one is needed. I think I know how to implement it and will try and finalize this tomorrow evening. If a list is last, back track until a non list is found and then just rebuild those templates you skipped traversing backwards.

Reply:
@svenevs it seems my code was way more fragile than I expected, it indeed was incorrect.
I've tested your tokenizer and it works way better! However, it seems to fail for nested templates because it can't go back up more than one level.
I have a proposed "fix": https://godbolt.org/z/bKW9z676T
IMO the tokenizer approach you added is 100% clearer and better, so we could just remove my broken code :P

Reply:
@florianhumblot awesome, it definitely took two brains for this one.  the initial implementation definitely had the right idea, just a few rough edges.  which mine did too xD

i've got some additional tests there that now pass with your fixed implementation

i will take another look at this soon, the joining of template names is not currently accurate (new helper function `join_template_tokens`) but it should be less bad than the tokenizing.

thanks so much for helping with this, it'll definitely merge just needs some more passes (slow turnaround for me)

Reply:
Thanks for the help @florianhumblot :heart:

I added your changes and some additional unit tests for them.  For now this conversation can be resolved!

I got hung up on some pesky details about `typename... Ts` or `class = void` in the templates, original tokenization with regex splitting up whitespace was no good.  I think we're good now but if you'd be willing to sanity check the implementation and just see if anything sticks out that would be great.

Reply:
The only template I can think of that might break the current tokenizer is this:
```c++
template<int (S<10>::*a)[10]>
struct S{};
``` 
That's `pointer to member object (of type int[10])` from the cppreference page you linked below.
It doesn't quite break, but the way it is tokenized makes me think it will easily break if we need to access the individual tokens:
```py
ipdb> >? tokenize_template("template<int (S<10>::*a)[10]>")
['template', ['int (S', ['10'], '::*a)[10]']]
```
Apart from that it looks like everything works as expected.

Reply:
It does look dubious... Still thinking about this one... Good eye

Reply:
I _think_ it's not going to be an issue, AFAICT that can only happen as a `< template parameter >` so even in the specialized case, it will be a list rather than a `str`.  So since the template join method still reconstructs it, it'll only ever show up in exhale if it's some specialization of something.

Or at least that's what I'm claiming, and if somebody ever comes along with code that breaks it then we can adapt.  But I'm definitely not interested in working on this anymore.